File: src/main/java/task/java/leetcode75/lvl1/IsomorphicStrings.java
package task.java.leetcode75.lvl1;

import java.util.HashMap;
import java.util.Map;

public class IsomorphicStrings {
    public boolean isIsomorphic(String s, String t) {
        if (s == null || t == null || s.length() != t.length())
            return false;
        if (checkValidRelation(t, s)) return false;

        if (checkValidRelation(s, t)) return false;

        return true;
    }

    private boolean checkValidRelation(String s, String t) {
        Map<Character, Character> map = new HashMap<>(t.length());
        for (int i = 0; i < t.length(); i++) {
            char key = s.charAt(i);
            char value = t.charAt(i);

            if (map.containsKey(key)) {
                if (!map.get(key).equals(value))
                    return true;
            }
            map.put(key, value);
        }
        return false;
    }
}

File: src/main/kotlin/task/kotlin/leetcode75/lvl1/IsomorphicStrings.kt
package task.kotlin.leetcode75.lvl1

// https://leetcode.com/problems/isomorphic-strings/
// time: O(n)
// space: O(n)

class IsomorphicStrings {
    fun isIsomorphic(s: String, t: String): Boolean {
        if (s.length != t.length) return false
        return check(s, t) && check(t, s)
    }

    private fun check(s: String, t: String): Boolean {
        val map = hashMapOf<Char, Char>()

        for (i in s.indices) {
            if (map.containsKey(s[i])) {
                if (map[s[i]] != t[i]) return false
            } else {
                map[s[i]] = t[i]
            }
        }
        return true
    }
}